File: js/heap.js
"use strict";

class MinHeap {
  constructor() {
    this.heap = [];
  }

  insert(value) {
    this.heap.push(value);
    this.bubbleUp(this.heap.length - 1);
  }

  extract() {
    if (this.heap.length <= 1) return this.heap.pop();

    const min = this.heap[0];
    this.heap[0] = this.heap.pop();
    this.sinkDown(0);
    return min;
  }

  bubbleUp(index) {
    const parentIndex = Math.floor((index - 1) / 2);
    if (index <= 0 || this.heap[parentIndex] <= this.heap[index]) return;

    [this.heap[index], this.heap[parentIndex]] = [this.heap[parentIndex], this.heap[index]];
    this.bubbleUp(parentIndex);
  }

  sinkDown(index) {
    const leftChildIndex = index * 2 + 1;
    const rightChildIndex = leftChildIndex + 1;
    let smallest = index;

    if (leftChildIndex < this.heap.length && this.heap[leftChildIndex] < this.heap[smallest]) {
      smallest = leftChildIndex;
    }

    if (rightChildIndex < this.heap.length && this.heap[rightChildIndex] < this.heap[smallest]) {
      smallest = rightChildIndex;
    }

    if (smallest !== index) {
      [this.heap[index], this.heap[smallest]] = [this.heap[smallest], this.heap[index]];
      this.sinkDown(smallest);
    }
  }
}

// Demonstrating usage
const minHeap = new MinHeap();
[8, 5, 1, 14, 9, 12, 13, 17, 15, 21].forEach(number => minHeap.insert(number));

console.log(`Heap before extracting min: [${minHeap.heap.join(", ")}]`);
for (let x = 0; x < 7; x++) {
  console.log(`Extracted min: ${minHeap.extract()}`);
  console.log(`Heap after extracting min: [${minHeap.heap.join(", ")}]`);
}

File: ruby/heap.rb
# frozen_string_literal: true

class MinHeap
  attr_accessor :heap

  def initialize
    @heap = []
  end

  def insert(value)
    @heap << value
    bubble_up(@heap.length - 1)
  end

  def extract
    return @heap.pop if @heap.length <= 1

    min = @heap[0]
    @heap[0] = @heap.pop
    sink_down(0)

    min
  end

  private

  def bubble_up(index)
    parent_index = (index - 1) / 2

    return if index <= 0 || @heap[parent_index] <= @heap[index]

    @heap[index], @heap[parent_index] = @heap[parent_index], @heap[index]
    bubble_up(parent_index)
  end

  def sink_down(index) # rubocop:disable Metrics/AbcSize
    left_child_index = (index * 2) + 1
    right_child_index = left_child_index + 1

    smallest = index
    smallest = left_child_index if left_child_index < @heap.length && @heap[left_child_index] < @heap[smallest]
    smallest = right_child_index if right_child_index < @heap.length && @heap[right_child_index] < @heap[smallest]

    return unless smallest != index

    @heap[index], @heap[smallest] = @heap[smallest], @heap[index]
    sink_down(smallest)
  end
end

min_heap = MinHeap.new
[8, 5, 1, 14, 9, 12, 13, 17, 15, 21].each do |number|
  min_heap.insert(number)
end

puts "Heap before extracting min #{min_heap.heap.compact.join(',')}"
7.times do
  puts "Extracted min: #{min_heap.extract}"
  puts "Heap after extracting min #{min_heap.heap.compact.join(',')}"
end
